core/src/main/java/com/yahoo/oak/EntryHashSet.java

```diff
@@ -0,0 +1,561 @@
+/*
+ * Copyright 2020, Verizon Media.
+ * Licensed under the terms of the Apache 2.0 license.
+ * Please see LICENSE file in the project root for terms.
+ */
+
+package com.yahoo.oak;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/* EntryHashSet keeps a set of entries placed according to the hash function.
+ * Entry is reference to key and value, both located off-heap.
+ * EntryHashSet provides access, updates and manipulation on each entry, provided its entry hash index.
+ *
+ * Entry is a set of at least 2 fields (consecutive longs), part of "entries" int array. The definition
+ * of each long is explained below. Also bits of each long (references) are represented in a very special
+ * way (also explained below). The bits manipulations are ReferenceCodec responsibility.
+ * Please update with care.
+ *
+ * Entries Array:
+ * --------------------------------------------------------------------------------------
+ * 0 | Key Reference          | Reference encoding all info needed to           |
+ *   |                        | access the key off-heap                         | entry with
+ * -----------------------------------------------------------------------------| entry hash index
+ * 1 | Value Reference        | Reference encoding all info needed to           | 0
+ *   |                        | access the key off-heap.                        |
+ *   |                        | The encoding of key and value can be different  |
+ * -----------------------------------------------------------------------------|
+ * 2 | Hash  - FULL entry hash index, which might be different (bigger) than    |
+ *   |         entry hash index                                                 |
+ * ---------------------------------------------------------------------------------------
+ * 3 | Key Reference          | Deleted reference (still) encoding all info     |
+ *   |                        | needed to access the key off-heap               |
+ * -----------------------------------------------------------------------------| Deleted entry with
+ * 4 | Value Reference        | Deleted reference(still)  encoding all info     | entry hash index
+ *   |                        | needed to access the key off-heap.              | 1
+ * -----------------------------------------------------------------------------|
+ * 5 | Hash  - FULL entry hash index, which might be different (bigger) than    |
+ *   |         entry hash index                                                 |
```

Comment:
I'm confused. Is a full hash simply the key's hash number? Or is it the hash+version? 
What is a "non-full-hash"?

I suggest calling the "hash+version" a "versioned hash".
The "full-hash" a "hash value".
And if there there is a use for a "non-full-hash", then maybe a "partial hash"? "hash subset"?
It depends on the use. I didn't find it yet :)

Reply:
Let's name the output of the hash function applied on the key a "hash number", which is integer: h(K) = hash number

So we have 3 issues here:
1. We have the lower X bits of the hash number serving as an index in the EntryArray. This is what I wanted to name "hash index".

2. We keep the hash number (all the bits) inside an entry. To separate it from the above I named it "full hash". I will refer to this name meanwhile in explanation of the issue 3 below.

3. The full hash is kept in a long entry field together with the update counter of this field. Sometimes it is required to read/update the entire "field" (full hash + update counter). But sometimes only the full hash needs to be extracted. I do not want to use word "version" here, because it is used for other purposes in Oak context. So I stick to "update counter".

As always the naming issue is the biggest problem in computer science, and the names I started with are not good enough. Let us decide about the names in the meeting. I am not going to fix it in the meanwhile.

Reply:
hashNumber = keyHash
hashIndex = index
getKeyHash & getKeyHashAndUpdateCounter

Reply:
Done